###################################################################################
#genome-mining to characterize the distribution and conservation of genes involved in the liberation and catabolism  
#of inorganic nitrogen across 6 genome-sequenced Suillus strains targeting Nitrate and Ammonia 

#inputs were generated using: 
#OrthoFinder 2.0
#trees were generated using Phyling/IQtree (geneme tree) or MAFFT/ClipKIT/IQtree ) (gene tree) 
#annotations were generated by JGI and linked to individual OGs using in house-scripts
#all associated code can be found on in the git repo https://github.com/MycoPunk/Suillus_N_use
#this script was run in R version 4.4.1
###################################################################################

#set packages 
library(msa)
library(ggmsa)
library(data.table)
library(tidyverse)
library(hrbrthemes)
library(forcats)
library(ggforce)
library(ggimage)
library(ggtree)
library(ape)
library(phytools)
library(ggplot2)
library(ggpubr)
library(quadprog)
library(devtools)
library(phangorn)
library(gridExtra)
library(circlize)
library(treeio)
library(ComplexHeatmap)
library(ComplexUpset)
sessionInfo()


##set seed for reproducibility
set.seed(666)

#Set variables 
n_Suillus_genomes= 6
setwd("~/Desktop/Project_N_use")

##read in data
OF.gene_families<-as.data.frame(fread("Orthogroups.tsv")) 
OF.unassigned<- as.data.frame(fread("Orthogroups_UnassignedGenes.tsv")) #this is where your singletons live
JGI_anno<- as.data.frame(fread("data/OG_linked_JGI_annotations.txt"), quote="") 
names(JGI_anno)<- c("Orthogroup", "species", "sseqid",	"iprId", "iprDesc",	"domainDb",	"domainId",	"domainDesc",	"numHits",	"domainStarts",	"domainEnds",	"score")
tree<- read.tree("Suillus.6_taxa.iqtree") #read in tree
tree_ammonium<-read.tree("data/ammonium_all_3_ref_alignment.aln.clipkit.treefile")
  
#BLASTP results for nitrate, in standard output format 6 w/ query length to calculate query coverage
NITRATE_HITS<- as.data.frame(fread("data/NITRATE.OUT.BLASTP_HITS_wqlen.aa.fasta", header = FALSE, sep = "\t")) #MEROPS blastp results
names(NITRATE_HITS)<- c("Orthogroup", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "qlen", "slen")
#add query coverage column
NITRATE_HITS$query_coverage <- (NITRATE_HITS$length / NITRATE_HITS$qlen) * 100
#round query coverage to nearest whole
NITRATE_HITS$query_coverage <- round(NITRATE_HITS$query_coverage, 0)
#add subject coverage column
NITRATE_HITS$subject_coverage <- (NITRATE_HITS$length / NITRATE_HITS$slen) * 100
#round subject coverage to nearest whole
NITRATE_HITS$subject_coverage <- round(NITRATE_HITS$subject_coverage, 0)


##subset results to only those that pass QC params -note this is already only displaying results with evalue <1.0e-10
#only those with query coverage > 70%
nrow(NITRATE_HITS) #9
NITRATE_HITS_qcov <- NITRATE_HITS[NITRATE_HITS$query_coverage >70, ]
nrow(NITRATE_HITS_qcov) #4
#only those with subject coverage > 70%
NITRATE_HITS_qcov <- NITRATE_HITS_qcov[NITRATE_HITS_qcov$subject_coverage >70, ]
nrow(NITRATE_HITS_qcov) #3
#only those with percent sequence identify >40%
NITRATE_HITS_qcov_pident <- NITRATE_HITS_qcov[NITRATE_HITS_qcov$pident >40, ]
nrow(NITRATE_HITS_qcov_pident) #3

##read in alignments for Nitrate
alignment_nitrate_reductase <- readAAMultipleAlignment("data/nitrate_reductase_alignment_w_ref", format="fasta")
alignment_nitrate_transporter <- readAAMultipleAlignment("data/nitrate_transporter_alignment_w_ref", format="fasta")
alignment_nitrite_reductase <- readAAMultipleAlignment("data/nitrite_reductase_alignment_w_ref", format="fasta")

##read in and process individual BLAST results for the OGs identified above 
OG0002188_transporter<- as.data.frame(fread("data/OG0002188_NITRATE.OUT.BLASTP_HITS", header = FALSE, sep = "\t")) #MEROPS blastp results
names(OG0002188_transporter)<- c("Orthogroup", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "qlen", "slen")
OG0002188_transporter$query_coverage <- (OG0002188_transporter$length / OG0002188_transporter$qlen) * 100
OG0002188_transporter$query_coverage <- round(OG0002188_transporter$query_coverage, 0)
OG0002188_transporter$subject_coverage <- (OG0002188_transporter$length / OG0002188_transporter$slen) * 100
OG0002188_transporter$subject_coverage <- round(OG0002188_transporter$subject_coverage, 0)
nrow(OG0002188_transporter)
OG0002188_transporter_qcov <- OG0002188_transporter[OG0002188_transporter$query_coverage >70, ]
nrow(OG0002188_transporter_qcov) #8
OG0002188_transporter_qcov <- OG0002188_transporter_qcov[OG0002188_transporter_qcov$subject_coverage >70, ]
nrow(OG0002188_transporter_qcov) #7
OG0002188_transporter_qcov_pident <- OG0002188_transporter_qcov[OG0002188_transporter_qcov$pident >40, ]
nrow(OG0002188_transporter_qcov_pident) #7


OG0002448_nitrate_reductase<- as.data.frame(fread("data/OG0002448_NITRATE.OUT.BLASTP_HITS", header = FALSE, sep = "\t")) #MEROPS blastp results
names(OG0002448_nitrate_reductase)<- c("Orthogroup", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "qlen", "slen")
OG0002448_nitrate_reductase$query_coverage <- (OG0002448_nitrate_reductase$length / OG0002448_nitrate_reductase$qlen) * 100
OG0002448_nitrate_reductase$query_coverage <- round(OG0002448_nitrate_reductase$query_coverage, 0)
OG0002448_nitrate_reductase$subject_coverage <- (OG0002448_nitrate_reductase$length / OG0002448_nitrate_reductase$slen) * 100
OG0002448_nitrate_reductase$subject_coverage <- round(OG0002448_nitrate_reductase$subject_coverage, 0)
nrow(OG0002448_nitrate_reductase) #8
OG0002448_nitrate_reductase_qcov <- OG0002448_nitrate_reductase[OG0002448_nitrate_reductase$query_coverage >70, ]
nrow(OG0002448_nitrate_reductase_qcov) #8
OG0002448_nitrate_reductase_qcov <- OG0002448_nitrate_reductase_qcov[OG0002448_nitrate_reductase_qcov$subject_coverage >70, ]
nrow(OG0002448_nitrate_reductase_qcov) #7 #lost one hit w/ low subject coverage
OG0002448_nitrate_reductase_qcov_pident <- OG0002448_nitrate_reductase_qcov[OG0002448_nitrate_reductase_qcov$pident >40, ]
nrow(OG0002448_nitrate_reductase_qcov_pident) #7


OG0003388_nitrite_reductase<- as.data.frame(fread("data/OG0003388_NITRATE.OUT.BLASTP_HITS", header = FALSE, sep = "\t")) #MEROPS blastp results
names(OG0003388_nitrite_reductase)<- c("Orthogroup", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "qlen", "slen")
OG0003388_nitrite_reductase$query_coverage <- (OG0003388_nitrite_reductase$length / OG0003388_nitrite_reductase$qlen) * 100
OG0003388_nitrite_reductase$query_coverage <- round(OG0003388_nitrite_reductase$query_coverage, 0)
OG0003388_nitrite_reductase$subject_coverage <- (OG0003388_nitrite_reductase$length / OG0003388_nitrite_reductase$slen) * 100
OG0003388_nitrite_reductase$subject_coverage <- round(OG0003388_nitrite_reductase$subject_coverage, 0)
nrow(OG0003388_nitrite_reductase) #8
OG0003388_nitrite_reductase_qcov <- OG0003388_nitrite_reductase[OG0003388_nitrite_reductase$query_coverage >70, ]
nrow(OG0003388_nitrite_reductase_qcov) #6 = lost two hits, both to the same doubled gene in S. americanus
OG0003388_nitrite_reductase_qcov <- OG0003388_nitrite_reductase_qcov[OG0003388_nitrite_reductase_qcov$subject_coverage >70, ]
nrow(OG0003388_nitrite_reductase_qcov) #6
OG0003388_nitrite_reductase_qcov_pident <- OG0003388_nitrite_reductase_qcov[OG0003388_nitrite_reductase_qcov$pident >40, ]
nrow(OG0003388_nitrite_reductase_qcov_pident) #6


##subset results to only those that pass QC params -note this is already only displaying results with evalue <1.0e-10
#only those with query coverage > 70%
nrow(NITRATE_HITS) #9
NITRATE_HITS_qcov <- NITRATE_HITS[NITRATE_HITS$query_coverage >70, ]
nrow(NITRATE_HITS_qcov) #4
#only those with percent sequence identify >40%
NITRATE_HITS_qcov_pident <- NITRATE_HITS_qcov[NITRATE_HITS_qcov$pident >40, ]
nrow(NITRATE_HITS_qcov_pident) #3



### AMMONIUM RESULTS ###

#BLASTP results for initial ammonium search using the longest representative for each OG, in standard output format 6 w/ query length to calculate query coverage
AMMONIUM_HITS<- as.data.frame(fread("data/AMMONIUM.3_ref_OUT.BLASTP_HITS.aa.fasta", header = FALSE, sep = "\t"))
names(AMMONIUM_HITS)<- c("Orthogroup", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "qlen", "slen")
#add query coverage column
AMMONIUM_HITS$query_coverage <- (AMMONIUM_HITS$length / AMMONIUM_HITS$qlen) * 100
#round query coverage to nearest whole
AMMONIUM_HITS$query_coverage <- round(AMMONIUM_HITS$query_coverage, 0)
#add subject coverage column
AMMONIUM_HITS$subject_coverage <- (AMMONIUM_HITS$length / AMMONIUM_HITS$slen) * 100
#round subject coverage to nearest whole
AMMONIUM_HITS$subject_coverage <- round(AMMONIUM_HITS$subject_coverage, 0)

##subset results to only those that pass QC params -note this is already only displaying results with evalue <1.0e-10
#only those with query coverage > 70%
nrow(AMMONIUM_HITS) #14
AMMONIUM_HITS_qcov <- AMMONIUM_HITS[AMMONIUM_HITS$query_coverage >70, ]
nrow(AMMONIUM_HITS_qcov) #12
#only those with subject coverage > 70%
AMMONIUM_HITS_qcov <- AMMONIUM_HITS_qcov[AMMONIUM_HITS_qcov$subject_coverage >70, ]
nrow(AMMONIUM_HITS_qcov) #9
#only those with percent sequence identify >40%
AMMONIUM_HITS_qcov_pident <- AMMONIUM_HITS_qcov[AMMONIUM_HITS_qcov$pident >40, ]
nrow(AMMONIUM_HITS_qcov_pident) #9

##read in alignments for Nitrate
alignment_ammonium_OG0009316 <- readAAMultipleAlignment("data/ammonium_OG0009316_alignment_w_ref.aln", format="fasta")
alignment_ammonium_OG0006295 <- readAAMultipleAlignment("data/ammonium_OG0006295_alignment_w_ref.aln", format="fasta")
alignment_ammonium_OG0006824 <- readAAMultipleAlignment("data/ammonium_OG0006824_alignment_w_ref.aln", format="fasta")


##read in and process individual BLAST results for the OGs identified above for ammonium 
OG0006824_ammonium_OG1<- as.data.frame(fread("data/OG0006824_NITRATE.OUT.BLASTP_HITS", header = FALSE, sep = "\t")) #MEROPS blastp results
names(OG0006824_ammonium_OG1)<- c("Orthogroup", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "qlen", "slen")
OG0006824_ammonium_OG1$query_coverage <- (OG0006824_ammonium_OG1$length / OG0006824_ammonium_OG1$qlen) * 100
OG0006824_ammonium_OG1$query_coverage <- round(OG0006824_ammonium_OG1$query_coverage, 0)
OG0006824_ammonium_OG1$subject_coverage <- (OG0006824_ammonium_OG1$length / OG0006824_ammonium_OG1$slen) * 100
OG0006824_ammonium_OG1$subject_coverage <- round(OG0006824_ammonium_OG1$subject_coverage, 0)
nrow(OG0006824_ammonium_OG1)#18
OG0006824_ammonium_OG1_qcov <- OG0006824_ammonium_OG1[OG0006824_ammonium_OG1$query_coverage >70, ]
nrow(OG0006824_ammonium_OG1_qcov) #18
OG0006824_ammonium_OG1_qcov <- OG0006824_ammonium_OG1_qcov[OG0006824_ammonium_OG1_qcov$subject_coverage >70, ]
nrow(OG0006824_ammonium_OG1_qcov) #18
OG0006824_ammonium_OG1_qcov_pident <- OG0006824_ammonium_OG1_qcov[OG0006824_ammonium_OG1_qcov$pident >40, ]
nrow(OG0006824_ammonium_OG1_qcov_pident) #18

##read in and process individual BLAST results for the OGs identified above for ammonium 
OG0006295_ammonium_OG2<- as.data.frame(fread("data/OG0006295_NITRATE.OUT.BLASTP_HITS", header = FALSE, sep = "\t")) #MEROPS blastp results
names(OG0006295_ammonium_OG2)<- c("Orthogroup", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "qlen", "slen")
OG0006295_ammonium_OG2$query_coverage <- (OG0006295_ammonium_OG2$length / OG0006295_ammonium_OG2$qlen) * 100
OG0006295_ammonium_OG2$query_coverage <- round(OG0006295_ammonium_OG2$query_coverage, 0)
OG0006295_ammonium_OG2$subject_coverage <- (OG0006295_ammonium_OG2$length / OG0006295_ammonium_OG2$slen) * 100
OG0006295_ammonium_OG2$subject_coverage <- round(OG0006295_ammonium_OG2$subject_coverage, 0)
nrow(OG0006295_ammonium_OG2)#18
OG0006295_ammonium_OG2_qcov <- OG0006295_ammonium_OG2[OG0006295_ammonium_OG2$query_coverage >70, ]
nrow(OG0006295_ammonium_OG2_qcov) #18
OG0006295_ammonium_OG2_qcov <- OG0006295_ammonium_OG2_qcov[OG0006295_ammonium_OG2_qcov$subject_coverage >70, ]
nrow(OG0006295_ammonium_OG2_qcov) #18
OG0006295_ammonium_OG2_qcov_pident <- OG0006295_ammonium_OG2_qcov[OG0006295_ammonium_OG2_qcov$pident >40, ]
nrow(OG0006295_ammonium_OG2_qcov_pident) #18


##read in and process individual BLAST results for the OGs identified above for ammonium 
#OG0009316_ammonium_OG3<- as.data.frame(fread("data/OG0009316_NITRATE.OUT.BLASTP_HITS", header = FALSE, sep = "\t")) #the 5 original seqs in the OG
OG0009316_ammonium_OG3<- as.data.frame(fread("data/OG0006824_NITRATE.OUT.BLASTP_HITS_w_truncated_seq", header = FALSE, sep = "\t")) #including the recovered truncated seq from Suiamp1
names(OG0009316_ammonium_OG3)<- c("Orthogroup", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "qlen", "slen")
OG0009316_ammonium_OG3$query_coverage <- (OG0009316_ammonium_OG3$length / OG0009316_ammonium_OG3$qlen) * 100
OG0009316_ammonium_OG3$query_coverage <- round(OG0009316_ammonium_OG3$query_coverage, 0)
OG0009316_ammonium_OG3$subject_coverage <- (OG0009316_ammonium_OG3$length / OG0009316_ammonium_OG3$slen) * 100
OG0009316_ammonium_OG3$subject_coverage <- round(OG0009316_ammonium_OG3$subject_coverage, 0)
nrow(OG0009316_ammonium_OG3) #18 (#15 w/ original OG)
OG0009316_ammonium_OG3_qcov <- OG0009316_ammonium_OG3[OG0009316_ammonium_OG3$query_coverage >70, ]
nrow(OG0009316_ammonium_OG3_qcov) #18 (#15 w/ original OG)
OG0009316_ammonium_OG3_qcov <- OG0009316_ammonium_OG3_qcov[OG0009316_ammonium_OG3_qcov$subject_coverage >70, ]
nrow(OG0009316_ammonium_OG3_qcov) #15 (#15 w/ original OG) - this is where we loose it, because of the truncation.
OG0009316_ammonium_OG3_qcov_pident <- OG0009316_ammonium_OG3_qcov[OG0009316_ammonium_OG3_qcov$pident >40, ]
nrow(OG0009316_ammonium_OG3_qcov_pident) #15 (#15 w/ original OG) #QC filtering did not remove any results- all very high matches. 



#mapping from JGI code to species names
names_df<- data.frame(cbind(JGI_code = c("Suicli1", "Suipic1", "Suiame1", "Suigr1", "Suilu4", "Suiamp1"), 
                            manuscript_name = c("S. clintonianus", "S. spraguei", "S. americanus", "S. weaverae", "S. luteus", "S. ampliporus")))


##clean and process data
#combine OF families with singletons
OF.gene_families_all<- rbind(OF.gene_families, OF.unassigned)

#clean up 
rm(OF.gene_families)
rm(OF.unassigned)

##basic stats:
#how many gene families are there?
n_gene_fams_all<- nrow(OF.gene_families_all)
n_gene_fams_all
#22,006


##fix names 
#remove .aa
names(OF.gene_families_all)<- sapply(names(OF.gene_families_all), gsub, pattern = ".aa", replacement = "" )
#create a named vector to rename with
rename_vector <- setNames(names_df$manuscript_name, names_df$JGI_code)
#rename the columns in OF.gene_families_all
OF.gene_families_all <- OF.gene_families_all %>% 
  rename_at(vars(names(rename_vector)), ~ rename_vector[.])


#how many strains?
#names of cols to exclude
cols_to_exclude<- "Orthogroup"

strains_only<- OF.gene_families_all[,!names(OF.gene_families_all) %in% cols_to_exclude]
strain_names<- names(strains_only)

#length(strain_names) check 
length(unique(strain_names)) 
#6


#add number_genomes column to get totals
#first replace blank calls with NAs
strains_only<- apply(strains_only, 2, function(x) gsub("^$|^ $", NA, x))
OF.gene_families_all$number_genomes_all<- rowSums(!is.na(strains_only)) #cols that are not blank

#replace blanks with NAs for all
OF.gene_families_all<- data.frame(apply(OF.gene_families_all, 2, function(x) gsub("^$|^ $", NA, x)), check.names=FALSE)

#change type to numeric
OF.gene_families_all$number_genomes_all <- as.numeric(as.character(OF.gene_families_all$number_genomes_all))


#how many of the gene families are in every genome including outgroups (of 6)?
n_gene_fams_core_all<- sum(OF.gene_families_all$number_genomes_all == n_Suillus_genomes)
n_gene_fams_core_all
#7,101 conserved in all 

#that's what % of the total?
(n_gene_fams_core_all*100)/n_gene_fams_all #32.27%

#how many of the gene families are singletons in Suillus
n_gene_fams_singletons<- sum(as.numeric(OF.gene_families_all$number_genomes_all) == 1)
n_gene_fams_singletons
#10,664

#that's what percent out of the total?
(n_gene_fams_singletons*100)/n_gene_fams_all
#48.46.13%




#### analysis of NITRATE genes ####

#subset BLAST P hits to only the top hit per OG
#NITRATE_top_hits<- NITRAE_HITS %>%
#  group_by(Orthogroup) %>%
#  slice(which.min(evalue)) %>%
#  ungroup()

#no need for the above, only one hit per OG. sanity check:

nrow(NITRATE_HITS_qcov_pident)
length(unique(NITRATE_HITS_qcov_pident$Orthogroup))

#by target what does the distribution of hits looks like?
table(NITRATE_HITS_qcov_pident$sseqid)

#1 OG for each: 
#nitrate reductase
#nitrate transporter
#nitrite reductase

#subset OGs to only those with BLAST hits to MEROPS database
NITRATE_subset<- semi_join(OF.gene_families_all, NITRATE_HITS_qcov_pident, by = "Orthogroup")

##remove annotations that didn't meet blast cut offs 
#get names of what was removed durring QC
OG0003388_removed <- OG0003388_nitrite_reductase[!(OG0003388_nitrite_reductase$Orthogroup %in% OG0003388_nitrite_reductase_qcov_pident$Orthogroup), "Orthogroup"]
OG0002448_removed <- OG0002448_nitrate_reductase[!(OG0002448_nitrate_reductase$Orthogroup %in% OG0002448_nitrate_reductase_qcov_pident$Orthogroup), "Orthogroup"]
OG0002188_removed <- OG0002188_transporter[!(OG0002188_transporter$Orthogroup %in% OG0002188_transporter_qcov_pident$Orthogroup), "Orthogroup"]

#remove the genes from the input dataframe
unique(OG0002188_removed) #Suiame1|Suiame1_1096338
unique(OG0002448_removed) #Suigr1|Suigr1_1012619
unique(OG0003388_removed) #Suiame1|Suiame1_986432
NITRATE_subset2<- NITRATE_subset
NITRATE_subset2[] <- lapply(NITRATE_subset2, function(x) gsub(", Suiame1\\|Suiame1_1096338", "", x))
NITRATE_subset2[] <- lapply(NITRATE_subset2, function(x) gsub(", Suigr1|\\Suigr1_1012619", "", x))
NITRATE_subset[] <- lapply(NITRATE_subset2, function(x) gsub(", Suiame1\\|Suiame1_986432", "", x))


#how many conserved across all 6?
#total_core<- sum(NITRATE_subset$number_genomes_all == 6)
#total_core #all six species are present in all three orthogroups


#### PLOT  ONTO TREE ####
#root tree
plot(tree)
ggtree(tree)

#root
tree<- ape::root(tree, "Suiamp1", resolve.root = TRUE,
                 interactive = FALSE, edgelabel = FALSE)

plot(tree)
ggtree(tree)
tree$tip.label

tree<- ladderize(tree, right = TRUE)
plot(tree)

#print nodes to see which one needs rotating
plot(tree)
nodelabels() #10 needs rotating
tree <- ape::rotate(tree, node = 10)
#check
tree$tip.label[tree$edge[tree$edge[, 2] <= length(tree$tip.label), 2]]

plot(tree)
ggtree(tree)

##rename the tips to match the rest of the manuscript
#function to rename tips
rename_tips <- function(tree, names_df) {
  #make a named vector for renaming
  rename_vector <- setNames(names_df$manuscript_name, names_df$JGI_code)
  #rename the tips
  tree$tip.label <- sapply(tree$tip.label, function(x) {
    if (x %in% names(rename_vector)) {
      return(rename_vector[x])
    } else {
      return(x)
    }
  })
  
  return(tree)
}

#run rename function
tree <- rename_tips(tree, names_df)

#print the leaves in the order they appear (starting at the root) with new names
tip_order <- tree$tip.label[tree$edge[tree$edge[, 2] <= length(tree$tip.label), 2]]
tip_order


#ggtree tip order
p <- ggtree(tree)

#extract the tip labels in the order they appear in the ggtree plot
tip_order <- p$data %>%
  filter(isTip) %>%
  arrange(y) %>%   # Arrange by y-axis position which corresponds to the vertical order in the plot
  pull(label)


###end tree stuff

##translate from gene names into gene counts
#first remove the last col with the totals
OF.gene_families_all_no_counts<- NITRATE_subset[, !(names(NITRATE_subset) %in% "number_genomes_all")]
#convert to counts
OF.gene_families_counts <- OF.gene_families_all_no_counts %>%
  mutate(across(-Orthogroup, ~ifelse(is.na(.), NA, as.character(str_count(., ",") + 1))))

#link the target for annotation of the clusters
OF.gene_families_counts_anno <- OF.gene_families_counts %>%
  inner_join(NITRATE_HITS %>% select(Orthogroup, sseqid), by = "Orthogroup") %>%
  select(Orthogroup, sseqid, everything()) #to set sseqid as the second column 

#reformat the ssqid column to just the names of the gene targets
OF.gene_families_counts_anno$sseqid <- OF.gene_families_counts_anno$sseqid %>%
  str_replace("([^_]+)_([^_]+)_.*", "\\1 \\2")


###out of curiosity, link the JGI annotations and see what they say -annotations from JGI using consecutive OG mapping starting with the longest transcript in the OG
how_informative_are_JGI_annos <- OF.gene_families_counts_anno %>%
  inner_join(JGI_anno %>% select(Orthogroup, iprDesc), by = "Orthogroup") %>%
  select(Orthogroup, sseqid, iprDesc, everything())
#interesting - mostly as expected.



###REFORMAT TO CAPTURE THE FUNCTION-LEVEL COUNTS FOR GRAPHING###

#remove OG col
OF.gene_families_counts_anno_clean<- subset(OF.gene_families_counts_anno[, -1])

#convert to numeric - except the annotation col
OF.gene_families_counts_anno_clean[, -1] <- lapply(OF.gene_families_counts_anno_clean[, -1], function(x) as.numeric(as.character(x)))
str(OF.gene_families_counts_anno_clean)

#set row names to the annotation 
rownames(OF.gene_families_counts_anno_clean) <- OF.gene_families_counts_anno_clean[, 1]
#and remove the col
OF.gene_families_counts_anno_clean <- OF.gene_families_counts_anno_clean %>%
  select(-sseqid)

#format for graphing
NITRATE<- t(OF.gene_families_counts_anno_clean) #this is your input for graphing. 



##graph all NITRATE genes##
#change datatype to matrix
NITRATE<- as.matrix(NITRATE)

#split each category to generate a color map
transporters <- NITRATE[, grep("nitrate transporter", colnames(NITRATE))]
nitrate_reductase <- NITRATE[, grep("nitrate reductase", colnames(NITRATE))]
nitrite_reductase <- NITRATE[, grep("nitrite reductase", colnames(NITRATE))]

#set color scale for each 
transporters_col_scale = colorRamp2(c(min(transporters)-2, max(transporters)), c("#FFFFFF", "#8C5F45"))
nitrate_reductase_col_scale = colorRamp2(c(min(nitrate_reductase)-2, max(nitrate_reductase))+.5, c("#FFFFFF", "#594D36"))
nitrite_reductase_col_scale = colorRamp2(c(min(nitrite_reductase)-2, max(nitrite_reductase))+1, c("#FFFFFF", "#8C5C03"))

#set names and order to graph in
NITRATE_names<- colnames(NITRATE)
desired_column_order <- c("nitrate transporter", "nitrate reductase", "nitrite reductase")

#Define the cell size
cell_size <- unit(7, "mm")  # Change this to your desired cell size

#find the correct row order based on tip_order
row_order <- rev(match(tip_order, rownames(NITRATE)))

#render the heatmap
nitrate_fig <- Heatmap(
  NITRATE, 
  name = "relative abundance \nper MEROPS family", 
  rect_gp = gpar(type = "none", col = "white", lwd = 2),  # Set white borders around each cell
  column_title = "NITRATE", 
  column_title_side = "bottom",
  column_names_rot = 45,
  cluster_columns = FALSE,  # Keep columns in the same order as the original data matrix
  show_row_dend = FALSE,  # Hide the row dendrogram
  show_column_dend = FALSE,  # Hide the column dendrogram
  row_order = row_order,  # Correctly set the row order
  row_names_gp = gpar(fontface = "italic"),
  top_annotation = HeatmapAnnotation(
    "total per \nMEROPS family" = anno_points(colSums(NITRATE)),
    height = unit(15, "mm")
  ),  
  right_annotation = rowAnnotation(
    "total per \nspecies" = anno_barplot(rowSums(NITRATE), width = unit(20, "mm"))
  ), 
  height = nrow(NITRATE) * cell_size,  # Set height of cells
  width = ncol(NITRATE) * cell_size,   # Set width of cells
  
  column_split = factor(colnames(NITRATE), levels = desired_column_order),
  column_order = order(factor(colnames(NITRATE), levels = desired_column_order)),  # Set the column order here
  
  cell_fun = function(j, i, x, y, width, height, fill) {
    v <- as.matrix(NITRATE)[i, j]
    
    # Define the color scaling function based on column names
    col <- ifelse(NITRATE_names[j] == "nitrate transporter", transporters_col_scale(v),
                  ifelse(NITRATE_names[j] == "nitrate reductase", nitrate_reductase_col_scale(v),
                         ifelse(NITRATE_names[j] == "nitrite reductase", nitrite_reductase_col_scale(v),
                                col_fun_c(v))))
    
    grid.rect(x, y, width, height, gp = gpar(fill = col, col = col))
    
    # Add text here
    grid.text(sprintf("%d", as.matrix(NITRATE)[i, j]), x, y, 
              gp = gpar(fontsize = 10, col = "black"))
  }, 
  show_heatmap_legend = FALSE
)

# Print the heatmap
nitrate_fig

pdf("nitrate_fig.pdf", width = 8, height = 4)
draw(nitrate_fig)
dev.off()



#### analysis of AMMONIUM genes ####

nrow(AMMONIUM_HITS_qcov_pident) #there are 9 results
length(unique(AMMONIUM_HITS_qcov_pident$Orthogroup)) #these exist across three different orthogroups: "OG0006824" "OG0006295" and "OG0009316"

#split to inspect 
OG0006824_initial_results<- AMMONIUM_HITS_qcov_pident[AMMONIUM_HITS_qcov_pident$Orthogroup == "OG0006824",] 
nrow(OG0006824_initial_results) 
length(unique(OG0006824_initial_results$sseqid)) #results for all! HcAmt1,2,3  
OG0006295_initial_results<- AMMONIUM_HITS_qcov_pident[AMMONIUM_HITS_qcov_pident$Orthogroup == "OG0006295",]
nrow(OG0006295_initial_results)
length(unique(OG0006295_initial_results$sseqid)) #results for all! HcAmt1,2,3 
OG0009316_initial_results<- AMMONIUM_HITS_qcov_pident[AMMONIUM_HITS_qcov_pident$Orthogroup == "OG0009316",]
nrow(OG0009316_initial_results)
length(unique(OG0009316_initial_results$sseqid)) #results for all! HcAmt1,2,3 

##BLASTP can not distinguish between the three-ish members of the superfmaily. Can we distinguish based on score?
#View(OG0006824_initial_results)
#OG0006824 lowest evalue is HcAmt3, highest bit score is HcAmt3 

#View(OG0009316_initial_results)
#OG0009316 lowest evalue is HcAmt3, highest bit score is HcAmt3 

#View(OG0006295_initial_results)
#OG0006295 lowest evalue is tied for HcAmt1/2, highest bit score is HcAmt1, but only by a small margin

#subset OGs to only those with BLAST hits to MEROPS database
AMMONIUM_subset<- semi_join(OF.gene_families_all, AMMONIUM_HITS_qcov_pident, by = "Orthogroup")

##translate from gene names into gene counts
#first remove the last col with the totals
OF.gene_families_all_no_counts<- AMMONIUM_subset[, !(names(AMMONIUM_subset) %in% "number_genomes_all")]
#convert to counts
OF.gene_families_counts <- OF.gene_families_all_no_counts %>%
  mutate(across(-Orthogroup, ~ifelse(is.na(.), NA, as.character(str_count(., ",") + 1))))

#convert to numeric - except the annotation (OG)col
OF.gene_families_counts[, -1] <- lapply(OF.gene_families_counts[, -1], function(x) as.numeric(as.character(x)))
str(OF.gene_families_counts)

#set row names to the annotation 
rownames(OF.gene_families_counts) <- OF.gene_families_counts[, 1]
#and remove the col
OF.gene_families_counts <- OF.gene_families_counts %>%
  select(-Orthogroup)

#format for graphing
AMMONIUM<- t(OF.gene_families_counts) #this is your input for graphing. 

#turn NA to 0
AMMONIUM[is.na(AMMONIUM)] <- 0


##make figure##
#split each category to generate a color map
OG0006824 <- AMMONIUM[, grep("OG0006824", colnames(AMMONIUM))]
OG0006295 <- AMMONIUM[, grep("OG0006295", colnames(AMMONIUM))]
OG0009316 <- AMMONIUM[, grep("OG0009316", colnames(AMMONIUM))]

#set color scale for each 
OG0006824_col_scale = colorRamp2(c(min(OG0006824)-1, max(OG0006824))-2, c("#FFFFFF", "#998A6D"))
OG0006295_col_scale = colorRamp2(c(min(OG0006295)-2, max(OG0006295))+.5, c("#FFFFFF", "#8C2C16"))
OG0009316_col_scale = colorRamp2(c(min(OG0009316)-1, max(OG0009316))+.5, c("#FFFFFF", "#D56035"))

#set names and order to graph in
AMMONIUM_names<- colnames(AMMONIUM)
desired_column_order <- c("OG0006295", "OG0006824", "OG0009316")

#Define the cell size
cell_size <- unit(7, "mm")  # Change this to your desired cell size

#find the correct row order based on tip_order
row_order <- rev(match(tip_order, rownames(AMMONIUM)))

#change datatype to matrix
AMMONIUM<- as.matrix(AMMONIUM)

#render the heatmap
ammonium_fig <- Heatmap(
  AMMONIUM, 
  name = "relative abundance \nper MEROPS family", 
  rect_gp = gpar(type = "none", col = "white", lwd = 2),  # Set white borders around each cell
  column_title = "AMMONIUM", 
  column_title_side = "bottom",
  column_names_rot = 45,
  cluster_columns = FALSE,  # Keep columns in the same order as the original data matrix
  show_row_dend = FALSE,  # Hide the row dendrogram
  show_column_dend = FALSE,  # Hide the column dendrogram
  row_order = row_order,  # Correctly set the row order
  row_names_gp = gpar(fontface = "italic"),
  top_annotation = HeatmapAnnotation(
    "total per \nMEROPS family" = anno_points(colSums(AMMONIUM)),
    height = unit(15, "mm")
  ),  
  right_annotation = rowAnnotation(
    "total per \nspecies" = anno_barplot(rowSums(AMMONIUM), width = unit(20, "mm"))
  ), 
  height = nrow(AMMONIUM) * cell_size,  # Set height of cells
  width = ncol(AMMONIUM) * cell_size,   # Set width of cells
  
  column_split = factor(colnames(AMMONIUM), levels = desired_column_order),
  column_order = order(factor(colnames(AMMONIUM), levels = desired_column_order)),  # Set the column order here
  
  cell_fun = function(j, i, x, y, width, height, fill) {
    v <- as.matrix(AMMONIUM)[i, j]
    
    # Define the color scaling function based on column names
    col <- ifelse(AMMONIUM_names[j] == "OG0006295", OG0006295_col_scale(v),
                  ifelse(AMMONIUM_names[j] == "OG0006824", OG0006824_col_scale(v),
                         ifelse(AMMONIUM_names[j] == "OG0009316", OG0009316_col_scale(v),
                                col_fun_c(v))))
    
    grid.rect(x, y, width, height, gp = gpar(fill = col, col = col))
    
    # Add text here
    grid.text(sprintf("%d", as.matrix(AMMONIUM)[i, j]), x, y, 
              gp = gpar(fontsize = 10, col = "black"))
  }, 
  show_heatmap_legend = FALSE
)

# Print the heatmap
ammonium_fig

#print out the plot
pdf("ammonium_fig.pdf", width = 8, height = 4)
draw(ammonium_fig)
dev.off()



### Make gene tree of AMT genes ###

# Rename the tips
names_df2<- data.frame(
  JGI_code = c("HcAmt1_AAM21926.1", "HcAmt2_AAK82416.1", "HcAmt3_AAK82417.1",
               "OG0006295_Suiame1_901902", "OG0006295_Suicli1_821153", "OG0006295_Suigr1_968494",
               "OG0006295_Suipic1_1473715", "OG0006295_Suiamp1_958622", "OG0006295_Suilu4_2925634",
               "OG0006824_Suiamp1_35551", "OG0006824_Suigr1_131374", "OG0006824_Suipic1_1476253",
               "OG0006824_Suiame1_930237", "OG0006824_Suicli1_816529", "OG0006824_Suilu4_2980670",
               "OG0009316_Suigr1_964606", "OG0009316_Suilu4_2845303", "OG0009316_Suiame1_899612",
               "OG0009316_Suicli1_71612", "OG0009316_Suipic1_1416813", 
               "missing_from_OG0009316_Suiamp1_1009242_gm1.16327_g"),
  manuscript_name = c("HcAmt1", "HcAmt2", "HcAmt3",
                      "OG0006295_S. americanus ID 901902", "OG0006295_S. clintonianus ID 821153", 
                      "OG0006295_S. weaverae ID 968494", "OG0006295_S. spraguei ID 1473715", 
                      "OG0006295_S. ampliporus ID 958622", "OG0006295_S. luteus ID 2925634",
                      "OG0006824_S. ampliporus ID 35551", "OG0006824_S. clintonianus ID 131374", 
                      "OG0006824_S. weaverae ID 1476253", "OG0006824_S. americanus ID 930237", 
                      "OG0006824_S. clintonianus ID 816529", "OG0006824_S. luteus ID 2980670",
                      "OG0009316_S. clintonianus ID 964606", "OG0009316_S. luteus ID 2845303", 
                      "OG0009316_S. americanus ID 899612", "OG0009316_S. clintonianus ID 71612", 
                      "OG0009316_S. weaverae ID 1416813", "OG0009316_S. ampliporus ID 1009242")
)



##rename the tips to match the rest of the manuscript
#run rename function
tree_ammonium <- rename_tips(tree_ammonium, names_df2)


#store the original tip labels before modifying them
original_labels <- tree_ammonium$tip.label

# Define the prefixes and their corresponding colors
prefix_colors <- list(
  "OG0006295" = "#8C2C16",
  "OG0006824" = "#998A6D",
  "OG0009316" = "#D56035"
)

#modify the tip labels to remove prefixes
tree_ammonium$tip.label <- str_replace(tree_ammonium$tip.label, "^(OG0006295_|OG0006824_|OG0009316_)", "")

#render tree
amt_tree <- ggtree(tree_ammonium, layout="rectangular") +
  geom_tiplab(size=2, hjust=-.2) +
  geom_nodelab(aes(label=label, x=x*1),
               size=2, 
               hjust=1, 
               vjust=0.5) +
  xlim(0, 5)  #increase xlim to make room for bars

#function to add clade label for a given OG  prefix
add_clade_label <- function(prefix, color, offset) {
  og_tips <- original_labels[grep(paste0("^", prefix, "_"), original_labels)]
  if(length(og_tips) > 0) {
    mrca_node <- getMRCA(tree_ammonium, str_replace(og_tips, paste0("^", prefix, "_"), ""))
    amt_tree <<- amt_tree + geom_cladelabel(node=mrca_node, label=prefix, color=color, 
                              offset=offset, 
                              offset.text=offset -.9,
                              barsize=2, angle=90, hjust=0.5)
  } else {
    print(paste("No tips with", prefix, "prefix found"))
  }
}

#add clade labels for each prefix
offsets <- seq(1, 1, length.out = length(prefix_colors))
for (i in seq_along(prefix_colors)) {
  add_clade_label(names(prefix_colors)[i], prefix_colors[[i]], offsets[i])
}

amt_tree

#print out the plot
ggsave("amt_tree_fig.pdf", plot = amt_tree, width = 8, height = 6)
